packages/hemi-tunnel-actions/src/finalizeWithdrawal.ts

```diff
@@ -0,0 +1,155 @@
+import { EventEmitter } from 'events'
+import {
+  Address,
+  Hash,
+  isAddress,
+  isHash,
+  PublicClient,
+  WalletClient,
+  publicActions,
+} from 'viem'
+import {
+  getWithdrawalStatus,
+  getWithdrawals,
+  publicActionsL1,
+  walletActionsL1,
+} from 'viem/op-stack'
+
+import { FinalizeEvents } from './types'
+import { toPromiseEvent } from './utils'
+
+type FinalizeWithdrawalParams = {
+  account: Address
+  l1WalletClient: WalletClient
+  l2PublicClient: PublicClient
+  withdrawalTransactionHash: Hash
+}
+
+const canFinalizeWithdrawal = async function ({
+  account,
+  l1WalletClient,
+  l2PublicClient,
+  withdrawalTransactionHash,
+}: FinalizeWithdrawalParams): Promise<{
+  canFinalize: boolean
+  reason?: string
+}> {
+  if (!isHash(withdrawalTransactionHash)) {
+    return { canFinalize: false, reason: 'invalid withdrawal transaction hash' }
+  }
+  if (!isAddress(account)) {
+    return { canFinalize: false, reason: 'account is not a valid address' }
+  }
+
+  const receipt = await l2PublicClient
+    .getTransactionReceipt({
+      hash: withdrawalTransactionHash,
+    })
+    .catch(() => null)
+
+  if (!receipt || receipt.status !== 'success') {
+    return {
+      canFinalize: false,
+      reason: 'Invalid or unsuccessful transaction receipt',
+    }
+  }
+
+  // Using @ts-expect-error fails to compile so I need to use @ts-ignore
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore because it works on IDE, and when building on its own, but fails when compiling from the portal through next
+  const withdrawalStatus = await getWithdrawalStatus(l1WalletClient, {
+    chain: l1WalletClient.chain,
+    receipt,
+    targetChain: l2PublicClient.chain,
+  })
```

Comment:
What if this call fails? It looks like the function could be more consistent if it always return an object with `canFinalize` and `reason` instead if also being able to throw. It makes the user have to consider more scenarios.

Reply:
Applied here [e870f3d](https://github.com/hemilabs/ui-monorepo/pull/1147/commits/e870f3dc65b0346a3faf5da703fa192fcb752d5c)